src/openrct2-ui/windows/Finances.cpp

```diff
@@ -326,46 +327,73 @@ static Widget _windowFinancesResearchWidgets[] =
-                            break;
-                        case WINDOW_FINANCES_PAGE_PROFIT_GRAPH:
-                            pageWidget = &_windowFinancesProfitWidgets[WIDX_PAGE_BACKGROUND];
-                            break;
-                        default:
-                            return;
-                    }
-                    _graphBounds = { windowPos + ScreenCoordsXY{ pageWidget->left + 4, pageWidget->top + 15 },
-                                     windowPos + ScreenCoordsXY{ pageWidget->right - 4, pageWidget->bottom - 4 } };
-                    _graphProps.RecalculateLayout(
-                        { _graphBounds.Point1 + kGraphTopLeftPadding, _graphBounds.Point2 - kGraphBottomRightPadding },
-                        kGraphNumYLabels, kGraphNumPoints);
-                    _graphProps.lineCol = colours[2];
-                }
-                break;
+                    graphPageWidget = &_windowFinancesCashWidgets[WIDX_PAGE_BACKGROUND];
+                    centredGraph = true;
+                    _graphProps.series = GetGameState().CashHistory;
+                    break;
+            }
+            // Recalculate size for graph pages only.
+            // Calculate Y axis max and min.
+            money64 maxVal = 0;
+            const auto series = _graphProps.series;
+            for (int32_t i = 0; i < kGraphNumPoints; i++)
+            {
+                auto val = std::abs(series[i]);
+                if (val == kMoney64Undefined)
+                    continue;
+                if (val > maxVal)
+                    maxVal = val;
             }
+            // This algorithm increments the leading digit of the max and sets all other digits to zero.
+            // e.g. 681 => 700.
+            money64 oom = 10;
+            while (maxVal / oom >= 10)
+                oom *= 10;
+            const money64 max = std::max(10.00_GBP, ((maxVal + oom - 1) / oom) * oom);
+
+            _graphProps.min = centredGraph ? -max : 0.00_GBP;
+            _graphProps.max = max;
+
+            // dynamic padding for long axis lables:
+            char buffer[64]{};
+            FormatStringToBuffer(buffer, sizeof(buffer), "{BLACK}{CURRENCY2DP} -", centredGraph ? -max : max);
+            int32_t maxWidth = GfxGetStringWidth(buffer, FontStyle::Small) + 1;
+            const ScreenCoordsXY dynamicPadding{ std::max(maxWidth, kGraphTopLeftPadding.x), kGraphTopLeftPadding.y };
+
+            _graphBounds = { windowPos + ScreenCoordsXY{ graphPageWidget->left + 4, graphPageWidget->top + 15 },
+                             windowPos + ScreenCoordsXY{ graphPageWidget->right - 4, graphPageWidget->bottom - 4 } };
+            _graphProps.RecalculateLayout(
+                { _graphBounds.Point1 + dynamicPadding, _graphBounds.Point2 - kGraphBottomRightPadding }, kGraphNumYLabels,
+                kGraphNumPoints);
+            _graphProps.lineCol = colours[2];
```

Comment:
I think it might be best to move this into a new function.

Reply:
I've extracted this to `OnPrepareDrawGraph`